fvm/meter/weighted_meter.go

```diff
@@ -308,6 +310,13 @@ func (params MeterParameters) MemoryWeights() ExecutionMemoryWeights {
 	return params.memoryWeights
 }
 
+// IsMemoryLimitSet returns if memory limit has been set by WithMemoryLimit()
+// It is useful when metering parameters are loaded from state. We need to
+// know if state has this value available to decide whether to override existing value.
+func (params MeterParameters) IsMemoryLimitSet() bool {
+	return params.isMemoryLimitSet
+}
+
 // TotalMemoryLimit returns the total memory limit
 func (params MeterParameters) TotalMemoryLimit() uint64 {
```

Comment:
This is not a safe pattern, because we assumes the caller will call `IsMemoryLimitSet` before `TotalMemoryLimit`. 
 It would cause a bug if the caller forget to call `IsMemoryLimitSet` before `TotalMemoryLimit`.

A safer change would be letting `TotalMemoryLimit` to return two values:
```
// TotalMemoryLimit returns the total memory limit
// it returns (0, false) if memory limit was not set
// it returns (limit, true) if memory limit was set by WithMemoryLimit()
func (params MeterParameters) TotalMemoryLimit() (uint64, bool) {
```

Reply:
Good point Leo. I thought about the use case of these interfaces, and considered it is fine to call `TotalMemoryLimit()` with or without `IsMemoryLimitSet()` since it will always returns valid value, even it is the default `math.MaxUint64`. For now the only scenario that `IsMemoryLimitSet()` is necessary is [here](https://github.com/onflow/flow-go/blob/master/fvm/executionParameters.go#L103), where we do need to know if state contains this setting to decide whether to override existing value. @janezpodhostnik

Reply:
In that case we don't need the `isMemoryLimitSet`, we could just do:

```
func (params MeterParameters) IsMemoryLimitSet() bool {
	return params.TotalMemoryLimit() == math.MaxUint64
}
```

Reply:
According to https://github.com/onflow/flow-go/blob/master/fvm/executionParameters.go#L103, we override existing meter setting only when state does contain this value, so if state does have `math.MaxUint64`, we still want to read it out and override the setting.

Reply:
OK. It's not clear to me where is `IsMemoryLimitSet` being used? 

> // We need to know if state has this value available to decide whether to override existing value.

How does this work? It seems we only set `isMemoryLimitSet`, never read from it, no?

Reply:
I have a draft PR (#3240) in which I plan to move out the state reading for meter params from fvm `RunV2()` which is called per tx, and do state reading at per collection level. the read state values will be passed into `RunV2()` in which this `IsMemoryLimitSet()` will be used to override existing setting values from `*Procedure`.

Reply:
I gave this a long think.

The problem is that the limit is coming from three sources:
- FVM hardcoded default
- context (set at node boot)
- state (controlled by governance)

the limit in the state can bu null at which point we should take the context one.

We currently do this  by doing:
```
meterParams := meter.DefaultParameters().
		WithComputationLimit(uint(proc.ComputationLimit(ctx))).
		WithMemoryLimit(proc.MemoryLimit(ctx))

	meterParams, err := getEnvironmentMeterParameters( 
		ctx,
		v,
		blockPrograms,
		meterParams, // <--- we only modify the limit if its actually set in the state. getEnvironmentMeterParameters already knows how to do this
	)
``` 

In PR https://github.com/onflow/flow-go/pull/3240/ you are moving `getEnvironmentMeterParameters` to the collection level, so the state parameters are read before the context ones and this causes that you need to keep track if the limit was set or not. 

I think the solution might be that the context read is also moved to the collection level. See my comment here: https://github.com/onflow/flow-go/pull/3240/files#r978947929

Reply:
that makes sense. so basically we should be move all meter setting logic out of `RunV2` and then we can get rid of this flag.